### PR TITLE
WIP: Feature/bpmn refactoring 2

### PR DESCRIPTION
**Pull request by dmytro-gundartsev**
### Motivation/Abstract

please describe in 1-3 sentences, 
* What is your motivation for this PR?
* What problem you try to solve, and how?

*Examples: This is a bug fix for ... ; I want to use this BPMN extension attributes 'X' for my project 'Y'...*

### Description/Comments

*please describe or comment, about your PR implementation choices, or what else you think is relevant for me to know*

### Checklist

Depending on your PR, please ensure the overall consistency/integrity of the project remains.
Please tick just one check item per section below

#### Tests
- [ ] did you update or create tests for your code changes?
- [ ] not relevant

#### Code examples
- [ ] did you update or add example code snippets, which relate to your code changes
- [ ] not relevant

#### Documentation
- [ ] did you update or create documentation, which relates to your code changes
- [ ] not relevant

### CODE REVIEW

File: pkg/bpmn/model/bpmn20/core.go [hunk 1]
```diff
@@ -93,41 +154,69 @@ type TFlowElement struct {
...
+	sourceRef           FlowNode
+	targetRef           FlowNode
+}
+
+type SequenceFlow interface {
+	FlowElement
+	GetSourceRef() FlowNode
+	GetTargetRef() FlowNode
+	GetConditionExpression() string
+	IsDefault() bool
```
Comment (richardbizik) [inline pkg/bpmn/model/bpmn20/core.go:202]: default is an attribute on the element itself (gateway) so I think we should remove it from here.

File: pkg/bpmn/model/bpmn20/core.go [hunk 2]
```diff
@@ -93,41 +154,69 @@ type TFlowElement struct {
...
+func (sequenceFlow *TSequenceFlow) GetTargetRef() FlowNode {
+	return sequenceFlow.targetRef
+}
+func (sequenceFlow *TSequenceFlow) GetConditionExpression() string {
+	// GetConditionExpression returns the embedded expression. There will be a panic thrown, in case none exists!
+	return html.UnescapeString(sequenceFlow.ConditionExpression.Text)
+}
+
+func (sequenceFlow *TSequenceFlow) IsDefault() bool {
+	return len(strings.TrimSpace(sequenceFlow.ConditionExpression.Text)) == 0
```
Comment (richardbizik) [inline pkg/bpmn/model/bpmn20/core.go:221]: default is specified by the preceding element not by presence of the expression in sequence flow.

File: pkg/bpmn/model/bpmn20/activities.go [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+	Implementation string `xml:"implementation,attr"`
+}
+
+type TUserTask struct {
+	TTask
+	// BPMN 2.0 Unorthodox elements. Part of the extensions elements
+	AssignmentDefinition extensions.TAssignmentDefinition `xml:"extensionElements>assignmentDefinition"`
+}
+
+func (task TTask) GetInputMapping() []extensions.TIoMapping {
```
Comment (richardbizik) [inline pkg/bpmn/model/bpmn20/activities.go:74]: I would move all the methods under its struct to keep them together

File: pkg/bpmn/model/bpmn20/core.go [hunk 4]
```diff
@@ -33,27 +39,81 @@ type TBaseElement struct {
...
+
+	for i := 0; i < val.NumField(); i++ {
+		fieldVal := val.Field(i)
+		if fieldVal.Kind() == reflect.Slice {
+			for j := 0; j < fieldVal.Len(); j++ {
+				arrEl := fieldVal.Index(j)
+				if !arrEl.CanInterface() || arrEl.Kind() != reflect.Struct {
+					continue
+				}
+				//aei := arrEl.Interface()
```
Comment (richardbizik) [inline pkg/bpmn/model/bpmn20/core.go:82]: clean these commented lines?

File: pkg/bpmn/model/bpmn20/core.go [hunk 5]
```diff
@@ -33,27 +39,81 @@ type TBaseElement struct {
...
+				var err = collectBaseElements(arrEl.Addr().Interface(), refs)
+				if err != nil {
+					return err
+				}
+			}
+		} else {
+			if !fieldVal.CanInterface() || fieldVal.Kind() != reflect.Struct {
+				continue
+			}
+			//fvi := fieldVal.Interface()
```
Comment (richardbizik) [inline pkg/bpmn/model/bpmn20/core.go:92]: clean these commented lines?

File: pkg/bpmn/model/bpmn20/core.go [hunk 6]
```diff
@@ -33,27 +39,81 @@ type TBaseElement struct {
...
+func (baseElement *TBaseElement) GetDocumentation() []Documentation {
 	var docs []Documentation
 	for _, doc := range baseElement.Documentation {
-		docs = append(docs, doc)
+		docs = append(docs, &doc)
 	}
 	return docs
 }
 
+func collectBaseElements(element interface{}, refs *map[string]BaseElement) error {
```
Comment (richardbizik) [inline pkg/bpmn/model/bpmn20/core.go:53]: I would like to see some tests for this where you build out more complicated struct in test itself and try to parse it.

File: pkg/bpmn/model/bpmn20/events.go [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+type TLinkEventDefinition struct {
+	Id   string `xml:"id,attr"`
+	Name string `xml:"name,attr"`
+}
+
+type TTimeDuration struct {
+	XMLText string `xml:",innerxml"`
+}
+
+func (startEvent TStartEvent) GetType() ElementType {
```
Comment (richardbizik) [inline pkg/bpmn/model/bpmn20/events.go:62]: move the methods next to their structs

File: pkg/bpmn/model/bpmn20/gateways.go [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+
+type GatewayElement interface {
+	FlowNode
+	IsParallel() bool
+	IsExclusive() bool
+	IsInclusive() bool
+}
+
+// Defaults
+func (parallelGateway TGateway) IsParallel() bool  { return false }
```
Comment (richardbizik) [inline pkg/bpmn/model/bpmn20/gateways.go:52]: move the methods next to their structs

File: pkg/bpmn/model/bpmn20/gateways.go [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+	GatewayDirection GatewayDirection `xml:"gatewayDirection,attr"`
+}
+
+const (
+	ElementTypeParallelGateway   ElementType = "PARALLEL_GATEWAY"
+	ElementTypeExclusiveGateway  ElementType = "EXCLUSIVE_GATEWAY"
+	ElementTypeEventBasedGateway ElementType = "EVENT_BASED_GATEWAY"
+	ElementTypeInclusiveGateway  ElementType = "INCLUSIVE_GATEWAY"
+
+	Unspecified GatewayDirection = "Unspecified"
```
Comment (richardbizik) [inline pkg/bpmn/model/bpmn20/gateways.go:17]: Can we rename these to have `GatewayDirection` prefix?

File: pkg/bpmn/model/bpmn20/helper_test.go [hunk 10]
```diff
@@ -9,25 +11,48 @@ import (
...
+	if err != nil {
+		t.Fatalf("failed to read file: %v", err)
+	}
+	err1 := xml.Unmarshal(xmlData, &definitions)
+	if err1 != nil {
+		t.Fatalf("failed to unmarshal XML: %v", err)
+	}
+	// Check that references in FlowNodes are correctly resolved
+	assert.Equal(t, "id", definitions.Process.ServiceTasks[0].GetId())
+	assert.Equal(t, &definitions.Process.SequenceFlows[0], definitions.Process.StartEvents[0].GetOutgoingAssociation()[0])
```
Comment (richardbizik) [inline pkg/bpmn/model/bpmn20/helper_test.go:48]: I would add check for expected length of these slices as well

**Review by richardbizik — CHANGES_REQUESTED** [review]

File: pkg/bpmn/model/bpmn20/core.go (hunk 1 above)
Comment (dmytro-gundartsev) [reply pkg/bpmn/model/bpmn20/core.go]: default is an optional attribute only on 3 gateways (inclusive and exclusive - now implemented + complex). Redesigned accordingly + fixed logic for inclusive and exclusive gateways (default is taken only when there is no conditional flows resolved). + Fixed the logic (with tests):
- In inclusive and exclusive gateways containing 2+ outgoing flows each flow has to have either expression or 'default' flag.
- In inclusive and exclusive gateways with 1 outgoing flow permitted the flow with no expression and no 'default flag'

File: pkg/bpmn/model/bpmn20/core.go (hunk 4 above)
Comment (dmytro-gundartsev) [reply pkg/bpmn/model/bpmn20/core.go:82]: Roger that

File: pkg/bpmn/model/bpmn20/gateways.go (hunk 8 above)
Comment (dmytro-gundartsev) [reply pkg/bpmn/model/bpmn20/gateways.go]: Roger that

File: pkg/bpmn/model/bpmn20/helper_test.go (hunk 10 above)
Comment (dmytro-gundartsev) [reply pkg/bpmn/model/bpmn20/helper_test.go:47]: Done

File: pkg/bpmn/model/bpmn20/gateways.go (hunk 9 above)
Comment (dmytro-gundartsev) [reply pkg/bpmn/model/bpmn20/gateways.go:17]: Done

File: pkg/bpmn/model/bpmn20/activities.go (hunk 3 above)
Comment (dmytro-gundartsev) [reply pkg/bpmn/model/bpmn20/activities.go]: Done

File: pkg/bpmn/model/bpmn20/core.go (hunk 6 above)
Comment (richardbizik) [reply pkg/bpmn/model/bpmn20/core.go:53]: It doesn't have to be a complex process...I meant that I would like to have a test that catches some corner cases ex: variable is/is not a pointer, variable is an interface (idk if we want to handle this but at least have a panic for that), array of pointer/direct vars etc...
